Add transducer-photoscan and photoscan-MRI ICP registration

sadhana-r opened the pull request:
Closes #325

Should be merged after #331. 

### For Review

This should be a quick review since it builds on the ICP logic introduced in #331. You can look over the code and test out ICP registration between the photoscan and transducer (final page of wizard).

arhowe00 reviewed (dismissed):
I tested out the ICP registration, and the transducer doesn't appear to be perfectly aligned with the photoscan. However, I don't doubt that I could have made mistakes in the process of aligning this, I just clicked through the wizard. Since this is to be merged after #331 is this behavior to be expected? I jumped on this review because it was requested recently but maybe it is too soon for this feature to work:
![image](https://github.com/user-attachments/assets/0f2edeb6-3e13-4528-88a6-b3a4435ed626)

Using the handles is nice, but I am not able to see the photoscan very well in demo_transducertracking. Is there a different subject/session you used in development that would be easier to use?
<img width="292" alt="image" src="https://github.com/user-attachments/assets/7738e688-d121-4080-abef-83464355785e" />

ebrahimebrahim commented:
Thanks for reviewing

> Is there a different subject/session you used in development that would be easier to use?

Yes, the Soren example is the one to try here.

> but maybe it is too soon for this feature to work:

No, it's supposed to work! However it is understood that transducer alignment only works if the initialization is close enough -- so it's okay if it's failing when the initialization is poor.

ebrahimebrahim commented:
Tests were very successful! For both photoscan-MRI and transducer-photoscan ICP, on the soren example.

Indeed one does have to give the transducer a good initialization for the ICP to work on the transducer.

The transducer registration surface could perhaps be improved here -- currently it's just a rectangle and it doesn't capture the sharp curvature at the edges of the outside surface of the device, which could be helpful for ICP.

ebrahimebrahim commented on OpenLIFUTransducerTracker/OpenLIFUTransducerTracker.py:
```diff
@@ -2121,6 +2212,156 @@ def run_fiducial_registration(self,
+
+            icp_result_node.SetMatrixTransformToParent( icpTransform.GetMatrix() )
+
+            if slicer.app.majorVersion >= 5 or (slicer.app.majorVersion >= 4 and slicer.app.minorVersion >= 11):
```
Woah, I think it's safe to assume we are on a least version 5 :sweat_smile: 

sooo many things would probably break if we tried any of this with Slicer 4...

sadhana-r replied:
True 😅 I copied whatever was in the logic from the Model Registration module but I'll remove this.

ebrahimebrahim commented on OpenLIFUTransducerTracker/OpenLIFUTransducerTracker.py:
```diff
@@ -2121,6 +2212,156 @@ def run_fiducial_registration(self,
+                icp_result_node.SetNodeReferenceID(slicer.vtkMRMLTransformNode.GetMovingNodeReferenceRole(), input_moving_model.GetID())
+                icp_result_node.SetNodeReferenceID(slicer.vtkMRMLTransformNode.GetFixedNodeReferenceRole(), input_fixed_model.GetID())
+
+        except Exception as e:
+            print("Exception thrown:", e)
+            return None
```
Arbitrary exceptions get swallowed up here and reduced to a print statement, but then in places where `run_icp_model_registration` is used we do not check if `None` was returned and do error handling.

I think it would be better to allow the exception to be raised, or re-raise it here. And ensure that if ICP fails then there is a reasonable user-facing outcome, not just a traceback or a print in the console. Could be as simple as an error dialog saying "ICP failed, check the logs."

ebrahimebrahim commented on OpenLIFUTransducerTracker/OpenLIFUTransducerTracker.py:
```diff
@@ -815,31 +878,56 @@ def setupTransformNode(self):
+        transducer_hardened.SetAndObserveTransformNodeID(self.transducer_to_volume_transform_node.GetID())
+        transducer_hardened.HardenTransform()
+ 
+        self.transducer_to_photoscan_icp_transform_node = self.wizard()._logic.run_icp_model_registration(
+            input_fixed_model = photoscan_hardened,
+            input_moving_model = transducer_hardened)
```
Should the transducer actually be registered using a _rigid_ transform (`transformType=0` here)?

Because the CAD model of the transducer comes with units and we use those to place it in the Slicer RAS space with mm as the units.

ebrahimebrahim replied:
(Update: Since they have asked for an early dev version of this to try out, I am going to make the change to `transformType=0` because I'm 80% sure that's what it should be -- but we can keep discussing when you get to this review)

sadhana-r replied:
Sounds good. You're right, the transducer registration should be a rigid transform. Going back to the point Peter brought up in our last meeting, since the photoscan is scaled to the skin segmentation, the photoscan transducer scale may be slightly different from the transducer scale so using a rigid transform here may not give the best fit right now, but it is the correct choice. 
For future reference, I think we need to maybe do a similarity transform between the photoscan and transducer to get the photoscan scaled correctly, and then rigid transform between the scaled photoscan and skin segmentation. Something like that..

ebrahimebrahim reviewed (changes requested):
A couple of small changes requested:

- Error handling in case ICP fails
- Consider whether transducer transform should be rigid

I looked over a lot of code, but not all of it -- it looks very good!

(Let me know if there is anything special you wanted to make sure I look at.)

By the way, this is a combined review of both #331 (which I closed without merge) and the additional things here.

And heads up -- I rebased this to main. So make sure to pull and reset before resuming on this branch.

Really wonderful work on this ICP integration!

sadhana-r commented:
> (Let me know if there is anything special you wanted to make sure I look at.)

Thanks for the review! The error handling was the main thing I wanted someone to look over so this was helpful! I have updated the exception handling now.

I also re-organized the commits introducing icp registration and squashed a couple commits so none of the commits introduce the Model Registration dependency.

ebrahimebrahim reviewed (approved):
Error handling looks good now!

> For future reference, I think we need to maybe do a similarity transform between the photoscan and transducer to get the photoscan scaled correctly, and then rigid transform between the scaled photoscan and skin segmentation. Something like that..

Agreed. Maybe an automated fixup step at the end that similarity-ICP-registers photoscan to transducer and then rigid-ICP-registers photoscan ROI to MRI again, and then finally again rigid-ICP-registers transducer to photoscan.

Or...

The MRI and the transducer both provide information that should help determine the scale of the photoscan. I wonder if we can come up with a way to simultaneously use both of them to optimally scale the photoscan?

Anyway, we can talk about that later

Merging!
